scvi/core/models/base.py

```diff
@@ -179,23 +180,41 @@ def save(self, dir_path: str, overwrite: bool = False):
                     dir_path
                 )
             )
-        torch.save(self.model.state_dict(), os.path.join(dir_path, "model_params.pt"))
-        with open(os.path.join(dir_path, "attr.pkl"), "wb") as f:
+
+        if save_anndata:
+            self.adata.write(os.path.join(dir_path, "adata.h5ad"))
+
+        model_save_path = os.path.join(dir_path, "model_params.pt")
+        attr_save_path = os.path.join(dir_path, "attr.pkl")
+        varnames_save_path = os.path.join(dir_path, "var_names.pkl")
+
+        var_names = self.adata.var_names
+        with open(varnames_save_path, "wb") as fp:
+            pickle.dump(var_names, fp)
```

Comment:
should we use h5 instead of pickle for this and the user attrs? I think you can put it in a dataframe and save hdf5 with pandas API. it might be cleaner

Reply:
I'd keep the model save file and the anndata files separate either way.

Reply:
I just played around with saving it as an h5. You need to cast the dtype of Index to save and theres no easy way to load and convert back to the user's dtype. It gets weird if you have like pd.Index(1, 2, '3') which pickle has no problem with.

Reply:
We could also just always cast to str and when checking, cast the 2nd adata.var_names to str as well and just compare like that. Edge case this will fail in is: pd.Index(0, '0')

Reply:
I think it's save to assume the var names are strs. regardless, you can maybe do instead `var_names.to_numpy()` and save using numpy's save method. Then you can make a new index on load from the numpy array and to check if they are equal, the index equals method.

Reply:
Okay, I'll just cast to str then and save as h5, since np.save just saves as pickle anyways.